[5주차] 유효성 검사

daeungkim opened the pull request:
안녕하세요! 이번주도 잘 부탁드립니다! ㅎㅎ
감사합니다.

daeungkim committed:
코드숨 5주차 강의 수강 및 반영

코드숨 5주차 강의 수강 및 반영

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -12,4 +16,9 @@ public String getGreeting() {
         SpringApplication.run(App.class, args);
     }
+
+    @Bean
```
`@Bean`을 사용하셨군요! 엄청 중요한 애노테이션이죠. `@Bean`에 대해 학습해서 역할과 기능을 간단히 설명해 보세요.

daeungkim replied:
먼저 Bean은 Spring IoC 컨테이너가 관리하는 자바 개체를 의미합니다.
IoC는 제어의 역전(IOC, Inversion Of Control)이라는 의미를 가지고 있습니다.
제어의 역전에서의 제어란, 개체를 생성하고 조작(메서드 호출 등)하는 작업을 의미합니다.
일반적인 프로그램에는 프로그래머가 모든 작업을 제어합니다.
IoC가 적용된 경우 개체의 생성을 특별한 관리 위임 주체에게 맡깁니다.
즉, 사용자의 제어권을 다른 주체에게 넘기는것을 IoC라고 합니다.


`@Bean`어노테이션는 어노테이션이 선언된 메서드가 Spring Container에게 Bean을 제공한다는 사실을 알려줍니다.
직접 제어 불가능한 외부 라이브러리등을 Bean으로 만들때 사용됩니다.
이와는 반대로 직접 작성한 클래스를 Bean으로 등록하기 위해서는 `@Component`를 사용합니다.

johngrib replied:
잘 조사하셨어요. 이번에는 `@Service`, `@Repository`와 `@Component`의 관계에 대해 조사해 보세요.

daeungkim replied:
`@Component`를 사용하여 개발자가 직접 작성한 클래스를 Bean으로 만들 수 있습니다.
`@Component`가 선언된 클래스는 `ClassPathBeanDefinitionScanner`에 의해 자동으로 스캔되어 등록됩니다.

`@Service`, `@Repository`, `@Controller`는 `@Component`의 특수한 타입입니다.

`@Repository`는 선언된 클래스가 data repository에 대해 정의한 것이라는 사실을 알려줍니다.
추가적으로 데이터베이스에서 발생하는 다양한 예외를 캐치하여 스프링의 DataAccessException으로 일괄 변환해 줍니다.

`@Controller`는 선언된 클래스가 컨트롤러의 역할을 하고 있음을 알려줍니다.
`@Controller`이 클래스에 선언된 경우 `@RequestMapping`이 선언된 메서드를 스캔하게 됩니다.

`@Service`는 어노테이션이 선언된 클래스가 어플리케이션 레이어에 해당함을 알려줍니다.

johngrib reviewed (commented):
오래간만입니다. 이번주도 화이팅입니다.

daeungkim committed:
Product 리소스관련 테스트에서 사용하는 데이터 정의

Product 리소스관련 테스트에서 사용하는 데이터 정의

daeungkim committed:
JpaProductRepository findAll 테스트 작성

JpaProductRepository findAll 테스트 작성
<description>

daeungkim committed:
Repository javadoc 작성

JpaProductRepository ProductRepository javadoc 작성

daeungkim committed:
ProductService getProducts 메서드 테스트 정의

ProductService getProducts 메서드 테스트 정의

daeungkim committed:
ProductService javadoc작성

ProductService getProducts 메서드 javadoc작성
<description>

daeungkim committed:
ProductController list메서드 테스트 작성

ProductController list메서드 테스트 작성

daeungkim committed:
ProductController list 메서드 javadoc 작성

ProductController list 메서드 javadoc 작성

daeungkim committed:
Product 테스트 코드 중복 해결

Product 테스트 코드 중복 해결

daeungkim committed:
"/products" GET 요청 통합 테스트 작성

"/products" GET 요청 통합 테스트 작성

daeungkim commented:
"/products" GET 요청 처리에 사용되는 클래스와 메서드 javadoc 및 테스트 작성하였습니다.

daeungkim commented on app/src/main/java/com/codesoom/assignment/domain/ProductRepository.java:
```diff
@@ -3,6 +3,10 @@
+/**
+ * Product 리소스
+ * Application layer에서 위임한 작업을 정의한다.
+ */
```
JpaProductRepository와 ProductRepository를 분리한 이유에 대해 잘 이해가 가지 않습니다.
이해가 깊지 않아서 ProductRepository 클래스에 대한 설명도 명확하지 않은것 같습니다.

Domain layer에서 db와의 통신을 책임지지 않도록 하기위해서 JpaProductRepository와 ProductRepository를 분리하였다고 이해하고 있는데요!
굳이 그렇게까지 해야하나 하는 느낌도 있고...
ProductRepository가 책임지는것이 무엇인지 잘 모르겠습니다.
혹시 참고할만한 자료나 검색 키워드 같은 것들이 있을까요?

johngrib replied:
제가 보기에도 굳이 그렇게까지 할 필요는 없어 보입니다. 지금 보기에는 둘 중 하나만 있어도 괜찮아 보입니다. 참고할만한 자료는 별로 없어요. 여기에서는 책임에 대해 최대한 열심히 생각해보시면 됩니다. 만약 책임을 나눠야 한다면 그 때 나눠도 늦지 않습니다.

daeungkim replied:
그렇군요!
고민해보겠습니다.
감사합니다. ㅎㅎ

daeungkim committed:
User 모델 정의

User 모델 정의

daeungkim committed:
UserRepository save 메서드 테스트 작성

UserRepository save 메서드 테스트 작성

daeungkim committed:
UserRepository JpaUserRepository 정의

UserRepository JpaUserRepository 정의

daeungkim committed:
ProductRepository javadoc 삭제

ProductRepository javadoc 삭제
-> ProductRepository 클래스의 책임에 대해 이해하지 못했음
-> 이해하지 못한 상태에서 작성한 javadoc이 해당 클래스 이해에 도움이 안된다고 판단

daeungkim committed:
JpaProductRepositoryTest 명칭 변경

JpaProductRepositoryTest ProductRepositoryTest로 명칭 변경
-> 코드에서 JpaProductRepository를 사용하지 않고 ProductRepository를 사용하므로

daeungkim committed:
Repository javadoc 수정 및 추가

Repository 클래스와 JpaRepository 클래스 javadoc 수정 및 추가
-> 도메인 계층은 엔티티 개체와 관련된 비즈니스 로직으로 구성되어 있음
-> Repository클래스는 비즈니스 로직을 정의하고
-> JpaRepository클래스는 비즈니스 로직과 데이터베이스를 연결하는 작업을 수행
-> 해당 내용이 잘 표현되도록 변경

daeungkim committed:
UserConstants user 데이터에서 id값 삭제

UserConstants user 데이터에서 id값 삭제
-> id를 사용하지 않기때문에 삭제하였음

daeungkim committed:
UserService createUser 메서드 테스트 작성

UserService createUser 메서드 테스트 작성

daeungkim committed:
UserService createUser 메서드 작성

UserService createUser 메서드 작성

daeungkim committed:
JSON 문자열 <-> 개체 간 변환 코드 ProductWebTest에서 분리

JSON 문자열 <-> 개체 간 변환 코드 ProductWebTest에서 분리
-> Controller 테스트에서도 사용하므로

daeungkim committed:
UserController create 메서드 테스트 작성

UserController create 메서드 테스트 작성

daeungkim committed:
User 엔티티 setter 삭제

User 엔티티 setter 삭제

daeungkim committed:
UserController create 메서드 작성

UserController create 메서드 작성

daeungkim committed:
UserController create 입력값이 잘못된 경우 테스트 추가

UserController create 입력값이 잘못된 경우 테스트 추가

daeungkim committed:
"/user" POST 요청 통합테스트 작성

"/user" POST 요청 통합테스트 작성

daeungkim committed:
"/products" GET 요청 통합 테스트 수정

@WebMvcTest 어노테이션 @SpringBootTest로 변경
-> @WebMvcTest의 경우 Service를 주입해주지 않음

daeungkim committed:
테스트 용도 Product ProductData 데이터 id 삭제

테스트 용도 Product ProductData 데이터 id 삭제
-> id 사용하지 않음
<description>

daeungkim committed:
MethodArgumentNotValidException 예외처리 코드 추가

MethodArgumentNotValidException 예외처리 코드 추가

daeungkim commented:
"/user" POST 요청 처리에 사용되는 코드, 관련 테스트 작성하였습니다.

johngrib commented on app/build.gradle:
```diff
@@ -51,6 +51,9 @@ dependencies {
     annotationProcessor 'org.projectlombok:lombok:1.18.16'
 
+    // DozerMapper
+    implementation 'com.github.dozermapper:dozer-core:6.4.0'
```
코드 리뷰의 형식으로 dozer mapper에 대해 소개해 주세요. 공식 문서 링크도 여기 달아주면 좋겠죠.

daeungkim replied:
매핑 프레임워크는 레이어드 아키텍쳐에서 유용합니다. 각각의 계층 이동시 데이터를 변환하여 사용할 수 있습니다.
Dozer mapper외에도 MapStruct ModelMapper 등이 있습니다.

```
Mapper mapper = DozerBeanMapperBuilder.buildDefault();
DestinationObject destObject = mapper.map(sourceObject, DestinationObject.class);
```
위의 예시처럼 소스에 해당하는 개체를 원하는 타입의 개체로 변환 할 수 있습니다.

```
public class SourceBean {

    private Long id;

    private String name;

    @Mapping("binaryData")
    private String data;

    @Mapping("pk")
    public Long getId() {
        return this.id;
    }

    public String getName() {
        return this.name;
    }
}
```
```
public class TargetBean {

    private String pk;

    private String name;

    private String binaryData;

    public void setPk(String pk) {
        this.pk = pk;
    }

    public void setName(String name) {
        this.name = name;
    }
}
```
위의 예시처럼 맴버변수의 이름 또는 데이터 타입이 다른경우 `@Mapping`어노테이션을 사용하여 해결할 수 있습니다.

현재 dozer mapper는 개발이 중단되었고, 곧 deprecated될 예정입니다.

레포지토리 주소 : https://github.com/DozerMapper/dozer
공식 사이트 주소 : https://dozermapper.github.io

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -4,20 +4,38 @@
 import javax.transaction.Transactional;
 import java.util.List;
 
+/**
+ * Product 리소스
+ * 수행할 작업을 정의하고, Domain layer에 수행을 위임한다.
+ *
+ * @see Product
+ * @see ProductRepository
+ */
```
이 정도면 충분하겠죠. `Domain layer에 수행을 위임한다.`처럼 설명을 적으면 마치 자신이 만화 속 인물인 것을 알고 있는 만화 주인공에게 자기 소개를 부탁한 것 같지 않나요?

```suggestion
/**
 * Product에 대한 조회, 입력, 수정 등을 담당합니다.
 */
```

한편 `@see ProductRepository`는 캡슐화를 깹니다. 클래스 javadoc은 클래스 바깥으로 공개되는 정보라고 생각하세요. 클래스 바깥으로 공개되는 정보에서 클래스 안쪽에서 몰래 사용하고 있는 `ProductRepository`를 드러낼 필요는 없습니다.

daeungkim replied:
아! 그렇군요! 수정하겠습니다. 감사합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,43 @@
+
+import org.springframework.stereotype.Service;
+
+/**
+ * User 리소스
+ * 수행할 작업을 정의하고, Domain layer에 수행을 위임한다.
+ *
+ * @see User
+ * @see UserRepository
+ */
```
여기도 수정해 보세요.

daeungkim replied:
넵! 수정하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/ArgumentNotValidAdvice.java:
```diff
@@ -0,0 +1,46 @@
+ */
+@ControllerAdvice
+public class ArgumentNotValidAdvice {
+    /**
+     * MethodArgumentNotValidException을 처리한다.
+     *
+     * @return 어디서(url경로, Http 메서드) 어떤 에러가 발생하였으며, 어떻게 해결할 수 있을지를 리턴한다.
+     *
+     * @see UserController#create(UserData)
+     *
+     */
```
이 정도로 충분하겠죠. `MethodArgumentNotValidException` 을 javadoc에 적는 것은 불필요하고, 커플링도 발생합니다. 예외 이름이 바뀌거나 다른 예외를 이 목적으로 쓰게 되었을 때 javadoc을 수정하지 않으면 잘못된 내용이 남게 되겠죠.

```suggestion
    /**
     * 던져진 "잘못된 메소드 입력값 예외"를 받아 에러 응답을 리턴합니다.
     *
     * @return 예외에 대한 내용이 담긴 응답
     */
```

한편 `@see UserController#create(UserData)`는 있어서는 안 되는 내용입니다. 이 예외를 던지는 곳이 어디어디인지 이 클래스에 알려져서는 안 됩니다. 얘는 공이 떠오르는 것을 기다리는 야구선수처럼 하늘만 보고 있다가 `MethodArgumentNotValidException`가 던져 올려지면 받는 역할만 하는 것입니다.

한편 예외 이름이 너무 포괄적이군요. 좀 더 구체적인 이름을 가진 예외 클래스를 사용하거나 적절히 번역해야 할 필요가 있어 보입니다.

daeungkim replied:
넵 수정하겠습니다.
예외 이름이 너무 포괄적이라는 말씀은 `MethodArgumentNotValidException`의 이름을 의미하는것인가요?
커스텀 예외를 만들어 사용하라는 의미로 이해하였는데, 제가 이해한것이 맞는지 궁금합니다.

johngrib replied:
맞아요. 김형욱님의 PR에 제가 남긴 코멘트도 함께 읽어보시면 도움이 되겠네요.
https://github.com/CodeSoom/spring-week5-assignment-1/pull/27#discussion_r705302968

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/NotFoundErrorAdvice.java:
```diff
@@ -8,12 +13,34 @@
+     *
+     * @return 어디서(url경로, Http 메서드) 어떤 에러가 발생하였으며, 어떻게 해결할 수 있을지를 리턴한다.
+     *
+     * @see ProductController#detail(Long)
+     * @see ProductService#getProduct(Long)
+     *
+     * @see ProductController#update(Long)
+     * @see ProductService#updateProduct(Long, ProductData)
+     *
+     * @see ProductController#destroy(Long)
+     * @see ProductService#deleteProduct(Long)
```
커플링이 발생하는 나쁜 주석이라 할 수 있습니다.
```suggestion
```

daeungkim replied:
넵 수정하겠습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/infra/UserRepositoryTest.java:
```diff
@@ -0,0 +1,49 @@
+
+    private User savedUser;
+
+    void subjectSave() {
+        savedUser = userRepository.save(USER);
+    }
```
`save` 메소드의 리턴값이 테스트 대상이죠? 그렇다면 이렇게 해 줍시다.

```suggestion
    User subjectSave() {
        return userRepository.save(USER);
    }
```

daeungkim replied:
넵 수정하겠습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/infra/UserRepositoryTest.java:
```diff
@@ -0,0 +1,49 @@
+        @Test
+        @DisplayName("User를 저장한다.")
+        void it_saves_object() {
+            subjectSave();
```
이렇게 받은 다음, 결과를 검증해보세요.
```suggestion
            User saveResult = subjectSave();
```

daeungkim replied:
넵 수정하겠습니다.

daeungkim replied:
```
void subjectDelete() {
    userRepository.delete(savedUser);
}
```
```
@AfterEach
void afterEach() {
         subjectDelete();
}
```
이 부분 때문에
```
void subjectSave() {
        savedUser = userRepository.save(USER);
}
```
```
@Test
@DisplayName("User를 저장한다.")
void it_saves_object() {
         subjectSave();
         assertThat(savedUser)
                  .matches(user -> user.getId() != null);
}
```
이런식으로 작성하였습니다.

johngrib replied:
subject 가 테스트하려는 메소드가 리턴값이 있다면, subject도 그것을 리턴하게 해 주세요. subject는 테스트 대상을 아주 얇게 감싼 것이어야 좋습니다. 리턴하는 것을 틀어막으면 좋지 않습니다.

daeungkim replied:
그렇군요! 수정하겠습니다.

daeungkim committed:
Update app/src/main/java/com/codesoom/assignment/application/ProductService.java

Co-authored-by: JohnGrib <johngrib.lee@kurlycorp.com>

daeungkim committed:
UserService javadoc 수정

UserService javadoc @see어노테이션 삭제
-> UserService의 캡슐화를 깨기 때문
UserService 클래스 설명 변경
-> Applicaion Layer에 대한 설명이 아닌 UserService에 대한 설명으로 변경

daeungkim committed:
ProductService javadoc 수정

ProductService javadoc 수정
-> 표현 방식 일관되도록 수정

daeungkim committed:
ArgumentNotValidAdvice javadoc 수정

ArgumentNotValidAdvice @see 어노테이션 삭제
-> 커플링이 발생하므로
ArgumentNotValidAdvice javadoc에서 MethodArgumentNotValidException 삭제
-> 커플링이 발생하며, 예외 이름이 바뀌거나 다른 예외를 해당 목적으로 사용하는경우 문제가 발생

daeungkim committed:
Update app/src/main/java/com/codesoom/assignment/controllers/NotFoundErrorAdvice.java

Co-authored-by: JohnGrib <johngrib.lee@kurlycorp.com>

daeungkim committed:
Merge branch 'main' of https://github.com/daeungkim/spring-week5-assignment-1

daeungkim committed:
NotFoundErrorAdvice javadoc 수정

NotFoundErrorAdvice javadoc @see 어노테이션 삭제
-> 커플링이 발생하므로
NotFoundErrorAdvice handleNotFound 메서드 설명 변경
-> ProductNotFoundException을 그대로 사용하는것은 커플링이 발생하며
-> 이후 변경하였을 때, javadoc을 수정해주어야함

daeungkim committed:
Update app/src/test/java/com/codesoom/assignment/infra/UserRepositoryTest.java

Co-authored-by: JohnGrib <johngrib.lee@kurlycorp.com>

daeungkim committed:
Merge branch 'main' of https://github.com/daeungkim/spring-week5-assignment-1

daeungkim committed:
Update app/src/test/java/com/codesoom/assignment/infra/UserRepositoryTest.java

Co-authored-by: JohnGrib <johngrib.lee@kurlycorp.com>

daeungkim committed:
Merge branch 'main' of https://github.com/daeungkim/spring-week5-assignment-1

daeungkim commented:
조언해주신 부분 변경하였습니다.

daeungkim committed:
커밋 되돌림

커밋 되돌림
-> 834d3e1a9ddd0a7d93f5b8c4530acd451663f4ad
-> beee798ea4ebd6e81543b3cfaf1e86d14303ce1e
-> 테스트 코드가 제대로 작동하지 않음

daeungkim committed:
ProductController, UserController javadoc 수정

UserController ProductController javadoc @see 어노테이션 삭제
-> 커플링이 발생하므로
UserController ProductController javadoc 변경
-> 역할을 좀 더 구제척으로 명시

daeungkim committed:
ProductConstants UserConstants 패키지 변경

ProductConstants UserConstants 패키지 변경
-> 테스트에서 반복적으로 사용되는 데이터의 모음이므로 도메인에 해당하지 않는다.

daeungkim committed:
ErrorResponse builder 패턴으로 변경

ErrorResponse builder 패턴으로 변경
-> 생성자를 사용하는것 보다 어떤 값들이 필요한지 명확하게 보이게 변경하였음

daeungkim committed:
UserData ProductData setter builder 삭제

UserData ProductData setter builder 삭제
-> 실제로 사용하지 않으므로

daeungkim committed:
@ParameterizedTest 테스트 표시 내용 변경

@ParameterizedTest 테스트 표시 내용 변경
-> 어떤 데이터가 들어가는지 명시

daeungkim committed:
ArgumentNotValidAdvice NotFoundErrorAdvice 합침

ArgumentNotValidAdvice NotFoundErrorAdvice 하나로 합침
-> 목적(컨트롤러에서 발생한 예외를 처리한다.)이 같으므로 굳이 나뉘어 있을 필요가 없음

daeungkim commented:
* Controller javadoc 수정하였습니다.
* ArgumentNotValidAdvice NotFoundErrorAdvice 클래스 하나로 합쳤습니다.

daeungkim committed:
UserRepository delete 메서드 테스트 작성

UserRepository delete 메서드 테스트 작성

daeungkim committed:
ProductNotFoundException 클래스 명 변경

ProductNotFoundException 클래스 명 변경
-> User를 찾지 못한 경우에도 사용해야 하므로

daeungkim committed:
UserService delete 메서드 테스트 작성

UserService delete 메서드 테스트 작성

daeungkim committed:
UserService deleteUser 메서드 구현

UserService deleteUser 메서드 구현

daeungkim committed:
UserController destroy 메서드 테스트 작성

UserController destroy 메서드 테스트 작성

daeungkim committed:
UserController destroy 메서드 작성

UserController destroy 메서드 작성

daeungkim committed:
"/user/{id}" DELETE 요청 통합테스트 작성

"/user/{id}" DELETE 요청 통합테스트 작성

daeungkim commented:
"/user/{id}" DELETE 요청 관련 기능 및 테스트 구현하였습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -1,23 +1,37 @@
         this.productRepository = productRepository;
     }
 
+    /**
+     * Product 목록 조회하고 리턴한다.
+     *
+     * @return Product 목록
+     */
```
이 정도가 적절하겠죠.
```suggestion
    /**
     * 상품 목록을 조회해 리턴합니다.
     *
     * @return 상품 목록
     */
```

daeungkim replied:
넵 수정하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -59,6 +61,8 @@ public Product deleteProduct(Long id) {
         return productRepository.findById(id)
-                .orElseThrow(() -> new ProductNotFoundException(id));
+                .orElseThrow(
+                    () -> new NotFoundException(Product.class.getSimpleName())
```
클래스 이름을 예외에 실어 던져봐야 알 수 있는 것은 클래스 이름 뿐입니다. `id`를 남긴다면 디버깅하는 사람을 도와줄 수 있겠죠.

```suggestion
                    () -> new NotFoundException(String.format("id %d 에 해당하는 상품을 찾지 못했습니다.", id))
```

daeungkim replied:
아! 그렇군요!
수정하겠습니다.
감사합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,56 @@
+    private User findUser(final Long id) {
+        return userRepository.findById(id)
+            .orElseThrow(
+                () -> new NotFoundException(User.class.getSimpleName())
```
여기도 수정이 필요해 보입니다.

daeungkim replied:
넵 수정하겠습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    /**
+     * User 생성 요청을 Application layer에 전달한다.
```
여기에서 레이어를 언급하는 것은 묘한 일입니다. 마치 만화책 주인공이, "내가 가진 물건을 다음 페이지에 나오는 인물에게 줘야지"하는 것 같은 표현입니다. 우리가 집이나 상점, 회사를 철근 덩어리라고 부르지는 않죠.

daeungkim replied:
Controller의 책임을 Http 요청 처리로 이해하였습니다.
Http 요청이 들어온 경우 Application layer의 서비스들이 이해할수 있는 형태로 변환시켜주는것이 Controller의 역할로 이해하였습니다.
"Http 요청을 Application layer에 전달한다." 혹은 "Http 요청을 Application layer에 연결한다."라고 표현하였습니다.
표현이 어색한것과는 별개로 제가 이해한 Controller의 역할에 오류가 있는지 궁금합니다.

daeungkim replied:
만약 이해에 오류가 없다면, Application layer가 아닌 서비스 클래스 명을 직접적으로 언급하는것이
커플링에 문제가 없을지 궁금합니다.
`User 생성 요청을 UserService에 전달한다` / `User 생성 요청을 UserService에 연결한다.`
혹은
`User 생성 요청을 UserService createUser에 전달한다` / `User 생성 요청을 UserService createUser에 연결한다.`
등으로 표현하였을 때 문제가 없을지 궁금합니다.

johngrib replied:
네 가지 모두 좋지 않습니다. javadoc을 작성할 때 가장 유의해야 하는 것은 "과정"을 설명하지 않는 것이에요. "결과"에 집중해 설명하는 것이 좋은 javadoc을 작성하게 될 가능성을 높여줍니다. 이 메소드들이 어떤 "응답"을 하는지를 생각해서 작성해보세요. 지금 이 javadoc 주석들은 모두 과정과 수단에 몰두하고 있어요.

daeungkim replied:
`javadoc을 작성할 때 가장 유의해야 하는 것은 "과정"을 설명하지 않는 것이에요. "결과"에 집중해 설명하는 것이 좋은 javadoc을 작성하게 될 가능성을 높여줍니다.` 명심하도록 하겠습니다. 감사합니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/infra/UserRepositoryTest.java:
```diff
@@ -0,0 +1,74 @@
+        return userRepository.findById(savedUser.getId());
+    }
+
+    @Nested
+    @DisplayName("save 메서드는")
+    public class Context_user_save {
+        @AfterEach
+        public void afterEach() {
+            subjectDelete();
+        }
+
+        @Test
+        @DisplayName("User를 저장한다.")
+        public void it_saves_users() {
+            savedUser = subjectSave();
+
+            assertThat(savedUser)
+                .matches(user -> user.getId() != null);
+        }
+    }
```
테스트 설명 `User를 저장한다`와 검증 코드 `user.getId() != null`이 일치하지 않습니다. 둘은 완전히 다른 활동입니다. 설명을 고치거나 검증 코드를 고쳐야 합니다.

daeungkim replied:
그렇군요! 테스트 수정하도록 하겠습니다. 감사합니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/ProductServiceTest.java:
```diff
@@ -1,138 +1,80 @@
+        void afterEach() {
+            verify(productRepository).findAll();
+        }
+        @Nested
+        @DisplayName("저장된 Product가 있다면")
+        class Context_product_exist {
+            @BeforeEach
+            void beforeEach() {
+                mockSubject()
+                    .thenReturn(PRODUCT_LIST);
+            }
+            @Test
+            @DisplayName("Product 목록을 리턴한다.")
+            void it_returns_a_product_list() {
+                subject()
+                    .isNotEmpty();
+            }
+        }
```
`repository`에 뭔가 문제가 생겨서 정반대로 작동하는 버그가 생겼다고 합시다. 뭔가 저장된 게 있으면 `findAll()`을 호출하면 비어 있는 리스트가 리턴되는 문제에요. 이 테스트는 그런 치명적인 버그가 생겼을 때에도 통과할 겁니다. 그렇다면 이 테스트의 의미는 무엇일까요?

한편 또 생각해 봐야 할 것이 있습니다. `저장된 Product가 있다면`과 `mockSubject().thenReturn(PRODUCT_LIST)` 사이의 의미상 불일치에요.

daeungkim replied:
```
        @Nested
        @DisplayName("저장된 Product가 있다면")
        class Context_product_exist {
            @BeforeEach
            void beforeEach() {
                mockSubject()
                    .thenReturn(EMPTY_LIST);
                    // .thenReturn(PRODUCT_LIST);
            }
            @Test
            @DisplayName("Product 목록을 리턴한다.")
            void it_returns_a_product_list() {
                subject()
                    .isNotEmpty();
            }
        }
```
`repository`에서 비어있는 리스트를 호출하도록 변경후 테스트를 돌려보았습니다.
이 경우 테스트를 실패하는것을 확인하였습니다.
<img width="1440" alt="Screen Shot 2021-09-10 at 5 20 36 AM" src="https://user-images.githubusercontent.com/45804974/132757042-5ca83b03-3eec-4d54-9ecc-32f872973708.png">
혹시 제가 잘못 이해한 부분이 있는지 궁금합니다.

daeungkim replied:
`저장된 Product가 있다면`과 `mockSubject().thenReturn(PRODUCT_LIST)` 사이에는 의미상의 불일치가 있는것 같습니다.
해당 테스트의 목적은 ProductRepositpory의 findAll 메서드 호출 여부와 findAll메서드의 리턴값을 그대로 리턴하는지 여부 입니다.
테스트 설명을 수정하도록 하겠습니다.

daeungkim replied:
```
package com.codesoom.assignment.application;

import com.codesoom.assignment.domain.Product;
import com.codesoom.assignment.domain.ProductRepository;

import org.assertj.core.api.ListAssert;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.mockito.stubbing.OngoingStubbing;

import java.util.Arrays;
import java.util.List;

import static com.codesoom.assignment.constants.ProductConstants.PRODUCT_LIST;
import static com.codesoom.assignment.constants.ProductConstants.EMPTY_LIST;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@Nested
@DisplayName("ProductService 클래스")
@ExtendWith(MockitoExtension.class)
class ProductServiceTest {
    @InjectMocks
    private ProductService productService;

    @Mock
    private ProductRepository productRepository;

    @Nested
    @DisplayName("getProducts 메서드는")
    class Describe_getProducts {
        ListAssert<Product> subject() {
            return assertThat(productService.getProducts());
        }

        OngoingStubbing<List<Product>> mockSubject() {
            return when(productRepository.findAll());
        }

        @AfterEach
        void afterEach() {
            verify(productRepository).findAll();
        }

        @Nested
        @DisplayName("ProductRepository findAll 메서드의 리턴값을")
        class Context_product_exist {
            @BeforeEach
            void beforeEach() {
                mockSubject()
                    .thenReturn(PRODUCT_LIST);
            }

            @Test
            @DisplayName("그대로 리턴한다.")
            void it_returns_a_product_list() {
                subject()
                    .matches(
                        outputList -> Arrays.deepEquals(
                            PRODUCT_LIST.toArray(), outputList.toArray()
                        )
                    );
            }
        }

        // @Nested
        // @DisplayName("저장된 Product가 없다면")
        // class Context_product_empty {
        //     @BeforeEach
        //     void beforeEach() {
        //         mockSubject()
        //             .thenReturn(EMPTY_LIST);
        //     }

        //     @Test
        //     @DisplayName("빈 목록을 리턴한다.")
        //     void it_returns_a_empty_list() {
        //         subject()
        //             .isEmpty();
        //     }
        // }
    }
}
```

johngrib replied:
`repository`에 뭔가 문제가 생긴 것과 `mockSubject`에 뭔가 문제가 생긴 것은 다른 문제입니다. 둘을 혼동하면 곤란해요. 저건 `repository`에 문제가 생긴 것이 아니죠.

daeungkim replied:
아! 그렇군요 이해하였습니다.
mock 개체를 사용하는 것이 repository의 작동을 보장하지 않는다는 말씀이시군요.
무의식적으로 테스트 대상을 제외한 모든 요소들을 mock개체로 만들어야 한다고 생각했던것 같습니다.
고민해보겠습니다. 감사합니다.

daeungkim committed:
Update app/src/main/java/com/codesoom/assignment/application/ProductService.java

Co-authored-by: JohnGrib <johngrib.lee@kurlycorp.com>

daeungkim committed:
ProductService 클래스 설명 javadoc 수정

ProductService 클래스 설명 javadoc 수정
-> 코드 리뷰를 통하여 list 메서드에대한 설명을 수정하였음
-> 메서드의 javadoc과 ProductService 클래스의 javadoc에 통일감을 주기 위하여 변경

daeungkim committed:
NotFoundException 예외 메세지에 id값 추가

NotFoundException 예외 메세지에 id값 추가
-> 어떤 id에서 발생하였는지 도움이 될 수 있으므로

daeungkim committed:
UserRepository save 메서드 테스트 수정

UserRepository save 메서드 테스트 수정
-> 테스트 설명 "User를 저장한다."와 검증코드 user.getId() != null이
-> 일치하지 않다는 조언을 받았음
-> 테스트 설명과 일치하는 검증코드가 되도록 수정
<description>

daeungkim committed:
ProductService getProducts 메서드 테스트 변경

ProductService getProducts 메서드 테스트 변경
-> ProductServiceTest의 테스트 설명문 "저장된 Product가 있다면"과
-> @BeforeEach 코드 mockSubject().thenReturn(PRODUCT_LIST) 사이의 불일치 발생
-> ProductService getProducts 메서드 테스트의 목적은
-> ProductRepository findAll 메서드의 리턴값을 그대로 리턴하는지 여부이므로
-> 기존의 테스트 설명문은 테스트의 목적을 제대로 표현하지 않음

daeungkim commented on app/src/test/java/com/codesoom/assignment/application/ProductServiceTest.java:
```diff
@@ -39,42 +40,52 @@
+
         @Nested
-        @DisplayName("저장된 Product가 있다면")
+        @DisplayName("ProductRepository findAll 메서드의 리턴값을")
```
https://github.com/CodeSoom/spring-week5-assignment-1/pull/26#discussion_r705325323
조언해주신 부분을 참고하여 테스트 코드를 변경하였습니다.
기존의 테스트에서 많은 부분이 달라졌습니다.
혹시 제가 잘못 이해한것인지 궁금합니다.

daeungkim commented:
코드에 피드백 반영하였습니다.
감사합니다.

daeungkim committed:
ProductService 불필요한 컨텍스트 삭제

ProductService 불필요한 컨텍스트 삭제

daeungkim committed:
ProductRepository subject 메서드 수정

테스트 대상인 메서드가 리턴타입이 있는 경우 subjec도 리턴값을 리턴할 수 있도록 변경
-> subject는 테스트 대상을 아주 얇게 감싼것이어야 좋다는 조언을 받았음

daeungkim committed:
테스트 접근제어자 수정

@Nested 클래스와 @Test 메서드를 제외한 모든 멤버변수 및 메서드 private 으로 변경
-> 굳이 public이나 패키지 접근제어자로 설정할 필요 없다 판단하였음

daeungkim committed:
ProductController list 메서드 테스트 변경

ProductService getProducts 메서트의 리턴값 리턴여부에 집중한 테스트 코드로 변경
-> ProductService getProducts의 리턴값이 빈 리스트인지 값이 들어있는 리스트인지는
-> ProductController입장에서 중요하지 않음

daeungkim committed:
ProductController @WebMvcTest 어노테이션 삭제

ProductController @WebMvcTest 어노테이션 삭제 및 테스트 반영
-> WebMvcTest의 경우 테스트하려는 Controller의 메서드만 테스트 하지 않음
-> 예를들어 Http Status 코드 리턴값 또는 ControllerAdvice 등
-> 이러한 경우 유닛테스트의 목적에 맞지 않음
<description>

daeungkim committed:
ProductController WebMvcTest 테스트 추가

ProductConstantsWebTest e2e테스트로 변경
ProductController WebMvcTest 테스트 추가
-> ProductController가 의도대로 작동한다는 가정하에
-> 나머지 부분(Http Status 코드 리턴 또는 Http 프로토콜 형식에 맞는 데이터 리턴 등)을
-> 테스트하는 코드 필요하므로

daeungkim commented:
"/products" GET 요청 관련 테스트 수정하였습니다.

ProductRepository(@DataJpaTest) -> ProductService -> ProductController -> ProductController(@WebMvcTest) -> E2E 테스트 순서로 테스트 작성하였습니다.

ProductService, ProductController 테스트 https://github.com/CodeSoom/spring-week5-assignment-1/pull/26#discussion_r705692131 형식으로 변경하였습니다.
@WebMvcTest의 경우 ProductController를 mock 개체로 만들어 나머지 부분(Http 프로토콜에 맞는 리턴값 확인, @ControllerAdvice 작동 여부 확인)을 테스트 하는 코드입니다.

daeungkim committed:
UserWebTest UserE2ETest로 변경

UserWebTest UserE2ETest로 변경하였음
-> WebTest의 목적은 Controller를 모킹한 상태에서 나머지 부분을 점검하는것이므로

daeungkim committed:
UserRepository findById 메서드 테스트 작성

UserRepository findById 메서드 테스트 작성

daeungkim committed:
User update 메서드 테스트 작성

User update 메서드 테스트 작성

daeungkim committed:
User update 메서드 작성

User update 메서드 작성

daeungkim committed:
UserService updateUser 메서드 테스트 작성

UserService updateUser 메서드 테스트 작성

daeungkim committed:
UserService updateUser 메서드 작성

UserService updateUser 메서드 작성

daeungkim committed:
UserController updateUser 메서드 테스트 작성

UserController updateUser 메서드 테스트 작성

daeungkim committed:
UserController updateUser 메서드 작성

UserController updateUser 메서드 작성

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/User.java:
```diff
@@ -0,0 +1,27 @@
+
+@Entity
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
이 세 애노테이션의 사용이 부적절합니다. 세 애노테이션이 왜 여기에 있는 것이 좋지 않은지에 대해 조사해 보세요.

daeungkim replied:
혹시 어떤 키워드를 조사하여야 하는지 알려주실 수 있나요?
제가 검색을 잘 못하는것인지, 관련 자료를 찾기가 힘드네요;;

johngrib replied:
예를 들어 이렇게 검색해 볼 수 있겠죠.
https://www.google.com/search?q=allargsconstructor+%EB%8B%A8%EC%A0%90

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/UserRepository.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.Optional;
+
+/**
+ * User 리소스 비즈니스 로직을 정의한다.
+ */
+public interface UserRepository {
```
리포지토리는 영속성 계층에 접근하는 전략을 캡슐화합니다. 비즈니스 로직이라는 단어는 보통 서비스에 좀 더 어울리는 단어죠. 리포지토리에는 잘 사용하지 않습니다.

daeungkim replied:
그렇군요. 계속해서 Repository 클래스의 역할에 대해 javadoc으로 작성하려고 하는데 조금씩 빗나가네요;; ㅎㅎ
계속해서 Repository 클래스의 역할, 책임에 대해 고민해보고 javadoc으로 작성하는 시도를 해보겠습니다.

johngrib replied:
조금씩 빗나가는 게 생각보다 꽤 소중한 경험입니다. 리뷰어가 있을 때 최대한 많이 실패해 보세요!

johngrib commented on app/src/test/java/com/codesoom/assignment/infra/ProductRepositoryTest.java:
```diff
@@ -0,0 +1,90 @@
+            @AfterEach
+            void afterEach() {
+                subjectDelete();
+            }
```
이렇게 하는 것이 좀 더 좋습니다.
```suggestion
            @BeforeEach
            void beforeEach() {
                subjectDelete();
                subjectSave();
            }

            @AfterEach
            void afterEach() {
                subjectDelete();
            }
```

daeungkim replied:
그렇군요! 이전 테스트에서 저장한 데이터들이 남아있어서 생기는 문제를 방지할 수 있겠군요.
조언 감사합니다.

daeungkim committed:
UserController update 웹테스트 작성

UserController update 웹테스트 작성

johngrib commented on app/src/test/java/com/codesoom/assignment/web/ProductWebTest.java:
```diff
@@ -0,0 +1,100 @@
+        }
+
+        @Test
+        @DisplayName("ProductController list 메서드의 리턴값을 JSON 문자열로 변환하여 리턴한다.")
```
이 테스트에 `ProductController`가 등장하면 테스트가 구현 자체에 의존하게 되어 좋지 않습니다. 

```suggestion
        @DisplayName("전체 목록 정보를 JSON 문자열로 변환하여 리턴한다.")
```

daeungkim replied:
넵 수정하겠습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/infra/UserRepositoryTest.java:
```diff
@@ -0,0 +1,125 @@
+            subjectDelete();
+        }
+
+        @Test
+        @DisplayName("User를 저장한다.")
+        public void it_saves_users() {
+            user = subjectSave();
+
+            assertThat(subjectFindById())
+                .isPresent();
+        }
+    }
```
테스트를 잘 작성하셨어요. 하지만 좀 더 디테일을 향상시킬 수 있습니다. 전역변수 `user`의 사용을 최대한 줄이고 다음과 같은 모양으로 테스트할 수 있도록 수정해 보세요. 실험실을 확실히 격리시켜서 깨끗한 테스트를 만드는 것이라 할 수 있습니다. 지금 코드를 읽어보면 `subjectFindById()`에서 좋지 못한 냄새가 납니다. 어떤 아이디를 가져오는지 확인하려면 메소드 구현을 찾아 읽어야 하는 문제도 있지요.

```suggestion
        @Test
        @DisplayName("User를 저장한다.")
        public void it_saves_users() {
            final User user = subjectSave();

            assertThat(subjectFindById(user.getId()))
                .isPresent();
        }
    }
```

daeungkim replied:
테스트 대상을 완전히 격리시키는것이 subject 메서드의 용도인 것으로 생각하였습니다.
따라서 어떤 값도 리턴하지 않고, 어떤 인자도 받지 않도록 하였습니다.
조언을 듣고 보니, 테스트 대상의 메서드 시그네처를 변경하더라도, subject 메서드 시그네처가 변경될 일은 매우 적겠군요.
착각했던것 같습니다. 조언 감사합니다.

johngrib replied:
이해하시면서 이해한 과정을 설명해주시는 것이 대웅님의 좋은 습관인 것 같습니다.

테스트 대상을 `subject` 내에 캡슐화시켜서, 테스트 대상 메소드의 이름이나 시그니처가 바뀌어도 테스트 코드의 일부만 수정해도 되도록 하는 것이 `subject`의 목표 중 하나라 할 수 있습니다. 테스트의 중요한 목표 중 하나는 테스트 대상의 리턴값을 검증하는 것이니 리턴값이 자연스럽게 흘러나오도록 해주는 것이 바람직하죠.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,175 @@
+        }
+
+        @Nested
+        @DisplayName("UserService createUser메서드가 User를 생성하면")
```
이 컨텍스트는 좋지 않습니다. `create` 메소드에 입력되는 값이 컨텍스트입니다. `UserService`의 행위를 컨텍스트로 추정하는 것은 이 테스트의 목적을 해칠 수 있습니다. `UserService`는 `create`의 캡슐화된 숨겨진 상태라 할 수 있기 때문이에요.

daeungkim replied:
넵 수정하도록 하겠습니다. 감사합니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,273 @@
+@DisplayName("UserService 클래스")
+@ExtendWith(MockitoExtension.class)
+public class UserServiceTest {
+    @InjectMocks
+    private UserService userService;
+
+    @Mock
+    private UserRepository userRepository;
+
+    @Mock
+    private Mapper dozerMapper;
```
Mock 없이 테스트를 작성하는 훈련을 먼저 해보세요. 테스트 코드를 작성하는 내공을 길러줄 겁니다. mock은 편리하지만 어쩔 수 없는 경우에만 제한적으로 쓰도록 연습해 봅시다. `UserRepository`를 검증하는 테스트 코드가 있기 때문에 여기에서는 안심하고 `UserRepository`를 mock 없이 테스트 준비 코드에 동원해도 됩니다.

```suggestion
```

daeungkim replied:
그렇군요! 왜인지 모르겠지만, 테스트 대상을 제외한 모든것을 mock으로 만들어야한다고 착각한것 같습니다.
mock없이 테스트를 작성하는 연습을 해보겠습니다.

johngrib replied:
착각이라는 것을 빠르게 판단하셨어요. mock에 필요 이상으로 과도하게 의존하게 되면 테스트의 많은 부분이 mock에 의한 추측으로 작동하게 됩니다. 지금 만들고 있는 프로젝트는 아주아주 작은 프로젝트라 티가 나지 않지만, 몇백만줄 정도의 대규모 프로젝트나, 매우 높은 안전성이 필요한 소프트웨어를 개발할 때에는 위험할 수 있습니다. 

그러니 처음에 테스트 코드의 감각을 익힐 때에는 mock을 사용하지 않고 최대한 진짜에 가깝게 테스트를 작성하는 연습을 하는 것이 필요합니다. 그러다가 mock을 사용할 수 밖에 없는 상황이 되었을 때 제한적으로 mock을 사용하는 식으로 접근해보세요. 

예를 들어 일정 관리 애플리케이션을 개발하는데, 기상청 API를 호출해서 날씨 정보를 받아오는 서비스 클래스가 하나 있다고 합시다. 그런데 이 클래스를 테스트할 때 기상청 API 서버가 점검중이라면 테스트 코드가 실패하게 되겠죠? 한편 기상청 API 서버가 멀쩡하면 테스트 코드는 성공하게 되겠죠? 테스트 코드가 외부에 의존하기 때문에 생기는 문제인 것입니다. 이렇게 우리가 제어할 수 없는 우리 영역 바깥을 다루는 코드는 mock 으로 다루게 되면 기상청의 상태가 어떤 상태이건 간에 테스트를 통해 정말 테스트하고 싶었던 것을 검증할 수 있어요. (우리가 테스트하고 싶은 건 기상청 서버가 아니었을 테니까요.)

이렇게 제한적으로 mock을 사용하다보면 mock의 가치와 응용 방법에 대한 깨달음을 얻으실 수 있을 거에요. 사용 범위도 조금씩 늘어나게 될 겁니다. 어떤 경우엔 mock을 쓰는 것이 적절해 보이지 않지만 다양한 상황을 고려해 mock을 쓰는 것이 결과적으로 더 낫다는 트레이드 오프 문제에 대해서도 나름의 결론을 내릴 수 있는 개발자가 될 겁니다. 그 때를 목표로 두고 노력해 봅시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/infra/JpaProductRepository.java:
```diff
@@ -8,6 +8,12 @@
+/**
+ * Product 리소스 비즈니스 로직과 데이터베이스를 연결한다.
+ *
+ * @see Product
+ * @see ProductRepository
```
이 두 줄은 불필요합니다. 3줄만 아래로 내려가도 인터페이스 시그니처에 두 클래스가 명시되어 있어요. 두 클래스와의 연결을 끊을 때 주석도 같이 수정해줘야 한다는 문제도 있습니다.

daeungkim replied:
넵 삭제하겠습니다.

daeungkim committed:
ProductController javadoc 작성

ProductController javadoc 작성

daeungkim committed:
UserController javadoc 작성

UserController javadoc 작성

daeungkim committed:
Repository javadoc 삭제

Repository javadoc 삭제
-> 비즈니스로직이라는 표현은 Repository보다 Service에 주로 쓰인다는 조언을 받았음

daeungkim committed:
Update app/src/test/java/com/codesoom/assignment/web/ProductWebTest.java

Co-authored-by: JohnGrib <johngrib.lee@kurlycorp.com>

daeungkim committed:
Merge branch 'main' of https://github.com/daeungkim/spring-week5-assignment-1

daeungkim committed:
테스트 설명에서 의존성에 대해 드러내는 부분 삭제하였음

테스트 설명에서 의존성에 대해 드러내는 부분 삭제하였음

daeungkim committed:
불필요한 javadoc 삭제

불필요한 javadoc 삭제

daeungkim committed:
User 생성 DTO와 수정 DTO 분리

User 생성 DTO와 수정 DTO 분리
-> 생성과 수정에서 필요한 데이터가 다르다.

daeungkim committed:
CrossOrigin 어노테이션 추가

CrossOrigin 어노테이션 추가

daeungkim committed:
User 생성 수정에 사용되는 DTO 분리에 따른 테스트 코드 수정

User 생성 수정에 사용되는 DTO 분리에 따른 테스트 코드 수정

daeungkim committed:
lombok config 파일 추가

lombok config 파일 추가

daeungkim commented:
테스트 커버리지 100% 달성하지 못했습니다.
주간 회고를 통하여 기간내에 과제를 완료하지 못한 이유를 고민해봐야할것 같습니다.
5주차 과제를 통하여 다양한 것들을 시도해 보았고, 다양한 시행착오를 겪었던것 같습니다.
이번 한주 리뷰해주시느라 수고하셨습니다.
다다음주에 뵙겠습니다.

johngrib reviewed (approved):
한 주간 고생 많으셨습니다. 노력하시는 모습이 멋집니다. 1주일 후에 또 만나요~
